### DISCUSSION TITLE
How do I manage file uploads in the new Route handlers?

### DISCUSSION THREAD

**Discussion by gmotta-dev** [vercel/next.js#50165] [App Router]
Im trying to use the new Route Handlers together with formidable but Im getting this error

`TypeError: req.on is not a function`

**Comment by pavel-surzhenko** [comment]
Same problem...

**Comment by icyJoseph** [comment]
Hi,

You can use `await req.formData()`

For example, with this code you can upload files to Digital Ocean Spaces.

```ts
import { S3, PutObjectCommand } from "@aws-sdk/client-s3";
import { File } from "buffer";
import { NextResponse } from "next/server";

const s3Client = new S3({
  forcePathStyle: false,
  endpoint: process.env.SPACES_ENDPOINT!,
  region: "us-east-1",
  credentials: {
    accessKeyId: process.env.SPACES_KEY!,
    secretAccessKey: process.env.SPACES_SECRET!,
  },
});

export async function POST(req: Request) {
  try {
    const form = await req.formData();
    const file = form.get("file-input"); // -> has to have the same name given in the form data object (client side)

    if (!file) return NextResponse.json({ message: "failure" });

    const isFile = file instanceof File;

    if (!isFile) return NextResponse.json({ message: "failure" });

    const buffer = await file.arrayBuffer();

    const data = await s3Client.send(
      new PutObjectCommand({
        Bucket: process.env.SPACES_NAME,
        Key: "a-name-for-this-element",
        Body: Buffer.from(buffer),
      })
    );

    console.log(data);
    return NextResponse.json({ message: "success" });
  } catch (reason) {
    console.log(reason);
    return NextResponse.json({ message: "failure" });
  }
}

```

**Reply by tsuccar** [reply]
could you point to docs that show other functions of "request". I only found req.formData() here. After using req.formData(), it gives deprecation warning as well. Thx in advance.

**Reply by icyJoseph** [reply]
Deprecation warning? Could you share a screenshot of that?

The Request there is the Request from the Web API Request:

- https://nextjs.org/docs/app/building-your-application/routing/router-handlers#extended-nextrequest-and-nextresponse-apis 

- https://developer.mozilla.org/en-US/docs/Web/API/Request


Note that it is only for the app router. And that while undici is polyfilled by Next.js in Node 16, it is probably a good idea to use Node 18+

**Reply by tsuccar** [reply]
Thank you @icyJoseph. I have not been able to reproduce what I had observed earlier since I have made some changes. However,  I do have a related issue to this topic  that I would love for you eyeball it for me. I'm stuck for a couple of days.  I'm trying to use 'multer' in to parse the formData in App router mode.  I'm have opened a a ticket in discussion area [unable to upload](https://github.com/vercel/next.js/discussions/50702)

**Reply by tsuccar** [reply]
Update to deprecating warning message: It's complaining about the config below  but goes away I remove it.
```js
export const config = {
  api: {
    externalResolver: true,
    bodyParser: false, 
  },
};
```

<img width="1248" alt="Screenshot 2023-06-03 at 10 17 26 AM" src="https://github.com/vercel/next.js/assets/4976030/05f9c851-ce30-46cd-9a11-802fcaac3f55">

**Reply by icyJoseph** [reply]
Hi, I don't have my laptop this weekend, I'll take a look at the other discussion Monday morning :) 

And that warning is because in App Router, there's no such config object anyway. You can remove it. 

I wonder if you really need multer, since the request can already be consumed as form data, all you really want now is the file validation before you upload them, right?

**Reply by tsuccar** [reply]
Hello Joseph, Appreciate you taking sometime during your off-time to respond :) . As a novice coder, i was at a loss on how to either integrate or replace  'multer + next-connect'  with 'App Router'. Enjoy your weekend and talk to you later.

**Comment by Tommy1223454** [comment]
`await req.formData()` is definitely the right transport primitive here.

One thing I’d add, though: getting a `File` object is not really the same as validating/trusting the upload. In production flows I’d usually inspect the file server-side before pushing it to object storage.

I built an OSS package for that step in Node.js:
https://github.com/pompelmi/pompelmi

It helps catch things like content/extension mismatches, suspicious archive structures, and other risky upload signals before the file is persisted or promoted further in the pipeline.
